File: elixir/03.exs
defmodule Day03 do
  @count 1
  @nocount 0
  @turnon 1
  @turnoff -1

  # Read input as lines; concatenate to a single line
  defp read_input do
    File.read!(Path.expand("../inputs/input03.txt"))
    |> String.replace("\n", "___newline___")
  end

  # Parse all the valid "mul(\d,\d)" substrings out of the line
  defp re_parse(line) do
    Regex.scan(~r/(do\(\))|(don't\(\))|(mul\((\d{1,3}),(\d{1,3})\))/, line)
    |> Enum.map(fn [head | tail] ->
      cond do
        String.match?(head, ~r/mul/) -> [@count, String.to_integer(Enum.at(tail, 3)), String.to_integer(Enum.at(tail, 4))]
        String.match?(head, ~r/don't\(\)/) -> [@nocount, @turnoff, @nocount]
        String.match?(head, ~r/do\(\)/) -> [@nocount, @turnon, @nocount]
      end
    end)
  end

  @doc """
  Sum the sums of products of the lines after parsing out the "mul(\d,\d)" substrings
  """
  def part1 do
    read_input()
    |> re_parse
    |> Enum.map(fn parts -> parts
      |> Enum.reduce(fn x, acc -> x * acc end)
    end)
    |> Enum.sum
    |> IO.inspect(label: "P1")
  end

  @doc """
  Sum the sums of products of the lines taking into account the "mul(\d,\d)", "do()", and "don't()" substrings
  """
  def part2 do
    tracker = %{value: 0, counting: true}

    read_input()
    |> re_parse
    |> Enum.reduce(tracker, fn [a, b, c], acc ->
      cond do
        a == @nocount && b == @turnoff -> Map.update!(acc, :counting, fn _ -> false end)
        a == @nocount && b == @turnon -> Map.update!(acc, :counting, fn _ -> true end)
        a == @count && acc.counting -> Map.update!(acc, :value, fn _ -> acc.value + (b * c) end)
        a == @count && !acc.counting -> acc
      end
    end)
    |> then(fn x -> x.value end)
    |> IO.inspect(label: "P2")
  end
end

# P1: 174103751
# P2: 100411201

File: php/03.php
<?php

const COUNT = 1;
const NOCOUNT = 0;
const TURNON = 1;
const TURNOFF = -1;

# Read input as lines; concatenate to a single line
function read_input()
{
    $data = file('../inputs/input03.txt'); // array of line strings
    $str = implode("___newline___", $data);
    return $str;
}

# Parse all the valid "mul(\d,\d)" substrings out of the line
function re_parse($line)
{
    $re = "/(do\(\))|(don't\(\))|(mul\((\d{1,3}),(\d{1,3})\))/";
    preg_match_all($re, $line, $matches, PREG_SET_ORDER);

    $mapped = [];
    foreach ($matches as $match) {
        if (substr($match[0], 0, 3) == "mul") {
            $mapped[] = [COUNT, (int) $match[4], (int) $match[5]];
        }
        else if ($match[0] == "don't()") {
            $mapped[] = [NOCOUNT, TURNOFF, NOCOUNT];
        }
        else if ($match[0] == "do()") {
            $mapped[] = [NOCOUNT, TURNON, NOCOUNT];
        }
    }
    return $mapped;
}

# Sum the sums of products of the lines after parsing out the "mul(\d,\d)" substrings
function part1() {
    $mapped_matches = re_parse(read_input());
    $products = array_map(function ($match) {
        return array_reduce($match, function ($acc, $x) {
            return $acc * $x;
        }, 1);
    }, $mapped_matches);

    return array_reduce($products, function ($acc, $p) {
        return $acc + $p;
    }, 0);
}

# Sum the sums of products of the lines taking into account the "mul(\d,\d)", "do()", and "don't()" substrings
function part2() {
    $tracker = ["value" => 0, "counting" => true];

    $mapped_matches = re_parse(read_input());

    $tracker = array_reduce($mapped_matches, function ($acc, $arr) {
        if ($arr[0] == NOCOUNT && $arr[1] == TURNOFF) {
            $acc["counting"] = false;
        }
        else if ($arr[0] == NOCOUNT && $arr[1] == TURNON) {
            $acc["counting"] = true;
        }
        else if ($arr[0] == COUNT && $acc["counting"]) {
            $acc["value"] += $arr[1] * $arr[2];
        }
        return $acc;
    }, $tracker);

    return $tracker["value"];
}

$p1 = part1();
print "P1: $p1\n"; # 174103751
$p2 = part2();
print "P2: $p2\n"; # 100411201
